### PR TITLE
FEC-related tweaks

### PR DESCRIPTION
**Pull request by cmc333333**
This collects @micahsaul's PRs.

* Fixes an issue with inserting a new section (without a subpart) -- previously, this triggered a resort, making the sections out of order
* Allows amdpars like "5. Section 100.94 is added to subpart C to read as follows" (specifically "to subpart C" as an indicator of context)
* Allow notices to have no section-by-section analyses
* Allow internal citations to have the form 11 CFR 110.1 (i.e. to look like an external citation, but refer to the current reg). Also allows for sequences of those.

### CODE REVIEW

**Comment by grapesmoker** [comment]
woooooo

On Wed, Mar 18, 2015 at 11:54 AM, CM Lubinski <notifications@github.com>
wrote:

> This collects @micahsaul <https://github.com/micahsaul>'s PRs.
>
>    - Fixes an issue with inserting a new section (without a subpart) --
>    previously, this triggered a resort, making the sections out of order
>    - Allows amdpars like "5. Section 100.94 is added to subpart C to read
>    as follows" (specifically "to subpart C" as an indicator of context)
>    - Allow notices to have no section-by-section analyses
>    - Allow internal citations to have the form 11 CFR 110.1 (i.e. to look
>    like an external citation, but refer to the current reg). Also allows for
>    sequences of those.
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cfpb/regulations-parser/pull/265
> Commit Summary
>
>    - Extend context_certainty for doc# 2014-24661
>    - Get 11 CFR 110 to parse without crashing. Fix issues where lack of
>    section_by_section was causing keyerror and new sections where being added
>    as siblings of emptypart as opposed to children.
>    - .
>    - Merge pull request #1 from 18F/stop_fec_crashes
>    - Correctly recognize internal citations of the form XX CFR YY if XX
>    and YY are the title and part being parsed. Also drop them from external
>    citations.
>    - Correctly recognize internal citations of the form XX CFR YY if XX
>    and YY are the title and part being parsed. Also drop them from external
>    citations.
>    - Add support for multiple internal CFR references in a single line
>    - Clean up external citation filtering
>    - Reverting last commit. Tests failing and logic require was uglier
>    than this.
>    - Add test, clean up unused params, and delete some extra whitespace
>    - Add one more test
>    - Merge pull request #2 from 18F/fix_internal_cites
>
> File Changes
>
>    - *M* regparser/citations.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-0>
>    (39)
>    - *M* regparser/grammar/amdpar.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-1> (2)
>    - *M* regparser/grammar/atomic.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-2> (2)
>    - *M* regparser/grammar/unified.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-3>
>    (18)
>    - *M* regparser/layer/external_citations.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-4> (6)
>    - *M* regparser/layer/internal_citations.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-5> (9)
>    - *M* regparser/layer/section_by_section.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-6> (2)
>    - *M* regparser/notice/compiler.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-7> (5)
>    - *M* tests/external_citation_parse.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-8>
>    (11)
>    - *M* tests/grammar_amdpar_tests.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-9> (9)
>    - *M* tests/layer_internal_citations_tests.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-10>
>    (33)
>    - *M* tests/layer_section_by_section_tests.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-11>
>    (10)
>    - *M* tests/notice_compiler_tests.py
>    <https://github.com/cfpb/regulations-parser/pull/265/files#diff-12>
>    (15)
>
> Patch Links:
>
>    - https://github.com/cfpb/regulations-parser/pull/265.patch
>    - https://github.com/cfpb/regulations-parser/pull/265.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cfpb/regulations-parser/pull/265>.
>



-- 
http://www.google.com/profiles/grapesmoker

File: regparser/grammar/unified.py [hunk 1]
```diff
@@ -170,3 +170,21 @@ def appendix_section(match):
...
+
+internal_cfr_p = (
+    atomic.title
+    + Suppress("CFR")
+    + atomic.part
+    + Suppress('.')
+    + atomic.section
+    + Optional(depth1_p))
+
+multiple_cfr_p = (
```
Comment (khandelwal) [inline regparser/grammar/unified.py]: I think we should start putting examples above these expressions to indicate what they are trying to parse. The unit tests help, but it would be nice to have the documentation right there.

File: regparser/layer/external_citations.py [hunk 2]
```diff
@@ -40,6 +40,12 @@ def parse(self, text, parts=None):
         cm = defaultdict(list)
         citation_strings = {}
         for citation, start, end in parser.scanString(text):
+            # Discard citations of form XX CFR YY if XX and YY are the title
+            # and part being parsed
+            if (citation[0] == self.cfr_title and
+                citation[1] == 'CFR' and
+                citation[2] == parts[0]):
+                continue
```
Comment (khandelwal) [inline regparser/layer/external_citations.py]: Is there a way to do this without a 'continue' here?

File: tests/layer_section_by_section_tests.py [hunk 3]
```diff
@@ -144,3 +144,13 @@ def test_process_order(self):
              "publication_date": "2010-10-10",
              "fr_volume": 22,
              "fr_page": 7676}])
+
+    def test_no_section_by_section(self):
+        notice = {
```
Comment (khandelwal) [inline tests/layer_section_by_section_tests.py]: This wasn't done previously in the code base, but it might be useful to add comments to tests to indicate why we expect it to do the thing we expect it to do. So, in this case something like: "There are sometimes notices without section by section analysis in them - let's ensure we handle those gracefully"

File: tests/notice_compiler_tests.py [hunk 4]
```diff
@@ -768,6 +768,21 @@ def test_add_node_no_parent(self):
         self.assertNotEqual(None, parent)
         self.assertEqual(parent.text, '')
 
+    def test_add_node_dummy_subpart(self):
+        root = Node(label=['1'])
```
Comment (khandelwal) [inline tests/notice_compiler_tests.py]: A comment on this test would be useful.

**Comment by cmc333333** [comment]
I wonder if @khandelwal has a label maker. How else could he keep his house well documented? :D

**Comment by khandelwal** [comment]
pfft.

**Comment by cmc333333** [comment]
Updated per Shashank's comments

**Comment by grapesmoker** [comment]
Is there a reason why we haven't merged it yet? There's functionality in here that would be useful to me.

**Comment by KimberlyMunoz** [comment]
@DarthDippy Is this ready to merge?

**Comment by ascott1** [comment]
@grapesmoker if you've reviewed this, I say merge away!

**Comment by KimberlyMunoz** [comment]
🙌
